saetest/saetest.go

```diff
@@ -85,10 +86,28 @@ func (c *EventCollector[T]) Unsubscribe() error {
 }
 
 // WaitForAtLeast blocks until at least `n` events have been received.
-func (c *EventCollector[T]) WaitForAtLeast(n int) {
+func (c *EventCollector[T]) WaitForAtLeast(ctx context.Context, n int) error {
+	quit := make(chan struct{})
+	defer close(quit)
+	go func() {
+		select {
+		case <-ctx.Done():
+			// If there is another waiter then we might not wake our own loop,
+			// so Signal() is inappropriate. For that matter, so too is
+			// Broadcast(). If context awareness without spurious wakeups is
+			// important, see https://www.youtube.com/watch?v=5zXAHh5tJqQ.
+			c.cond.Broadcast()
+		case <-quit:
+		}
+	}()
+
 	c.cond.L.Lock()
 	defer c.cond.L.Unlock()
 	for len(c.all) < n {
 		c.cond.Wait()
+		if ctx.Err() != nil {
+			return context.Cause(ctx)
+		}
 	}
+	return nil
 }
```

Comment:
This implementation of `WaitForAtLeast` does not actually guarantee context cancellation will result in waking up the goroutine.

(Shameless self plug): I'd recommend using the `lock.Cond` implementation here: https://github.com/ava-labs/avalanchego/blob/master/utils/lock/cond.go which has the same usage as the `sync.Cond` implementation but changes `Wait` to take in a context and return an error.